### ISSUE TITLE
Optimizer changes for ST

### ISSUE THREAD

**Issue by paf31** [purescript/purescript#3100]
`ST` will no longer depend on `Eff` and so the inliner will need to change slightly.

**Comment by colinwahl** [comment]
I know this has been open for a while, but I'd like to give it a shot!

**Comment by matthewleon** [comment]
Is the new `ST` in 0.12 not being inlined? @colinwahl can you provide a sample of some compiled ST code?

**Comment by colinwahl** [comment]
Sure, give me a few minutes

**Comment by colinwahl** [comment]
Here is a function that uses ST:

```purescript
count :: Int
count = ST.run do
  c <- ST.new 0
  ST.for 0 11 \i -> do
    ST.modify (add i) c
  ST.read c
```

and here is the compiled function:
```purescript
var count = Control_Monad_ST_Internal.run(Control_Bind.bind(Control_Monad_ST_Internal.bindST)(Control_Monad_ST_Internal["new"](0))(function (v) {
    return Control_Bind.discard(Control_Bind.discardUnit)(Control_Monad_ST_Internal.bindST)(Control_Monad_ST_Internal["for"](0)(11)(function (i) {
        return Control_Monad_ST_Internal.modify(Data_Semiring.add(Data_Semiring.semiringInt)(i))(v);
    }))(function () {
        return Control_Monad_ST_Internal.read(v);
    });
}));
```

This is using 
```
Pulp version 12.3.0
purs version 0.12.0
```

**Comment by colinwahl** [comment]
I implemented `magicDo` for ST and immediately saw an improvement. I am now working on fixing `inlineST`.  You can track my work on this branch: https://github.com/colinwahl/purescript/tree/fix-st-inlining

**Comment by matthewleon** [comment]
Good to see this happening. Happy to pitch in if you get stuck at any point. I figure you can probably reuse a lot of the logic from the `Eff ST` inlining mechanism.

**Comment by natefaubion** [comment]
This was fixed in #3444
